bolt-modules/boltlib/lib/puppet/functions/apply_prep.rb

```diff
@@ -61,21 +83,47 @@ def apply_prep(target_spec)
         agent_targets.each { |target| Puppet.debug "Puppet Agent feature declared for #{target.name}" }
         unless unknown_targets.empty?
           # Ensure Puppet is installed
-          versions = run_task(executor, unknown_targets, 'puppet_agent::version')
+          version_task = get_task('puppet_agent::version')
+          versions = run_task(unknown_targets, version_task)
+          raise Bolt::RunFailure.new(versions, 'run_task', 'puppet_agent::version') unless versions.ok?
           need_install, installed = versions.partition { |r| r['version'].nil? }
           installed.each do |r|
             Puppet.debug "Puppet Agent #{r['version']} installed on #{r.target.name}"
-            inventory.set_feature(r.target, 'puppet-agent')
+            set_agent_feature(r.target)
           end
 
           unless need_install.empty?
             need_install_targets = need_install.map(&:target)
-            run_task(executor, need_install_targets, 'puppet_agent::install')
-            # Service task works best when targets have puppet-agent feature
-            need_install_targets.each { |target| inventory.set_feature(target, 'puppet-agent') }
-            # Ensure the Puppet service is stopped after new install
-            run_task(executor, need_install_targets, 'service', 'action' => 'stop', 'name' => 'puppet')
-            run_task(executor, need_install_targets, 'service', 'action' => 'disable', 'name' => 'puppet')
+            # lazy-load expensive gem code
+            require 'concurrent'
+            pool = Concurrent::ThreadPoolExecutor.new
+
+            hooks = need_install_targets.map do |t|
+              begin
+                opts = t.plugin_hooks&.fetch('puppet_library')
+                hook = inventory.plugins.get_hook(opts['plugin'], :puppet_library)
+                { 'target' => t,
+                  'hook_proc' => hook.call(opts, t, self) }
+              rescue Bolt::Error => e
+                Bolt::Result.from_exception(t, e)
+              end
+            end
+
+            hook_errors, ok_hooks = hooks.partition { |h| h.is_a?(Bolt::Result) }
+
+            futures = ok_hooks.map do |hash|
+              Concurrent::Future.execute(executor: pool) do
+                hash['hook_proc'].call
+              end
+            end
+
+            results = futures.zip(ok_hooks).map do |f, hash|
+              f.value || Bolt::Result.from_exception(hash['target'], f.reason)
+            end
+            set = Bolt::ResultSet.new(results + hook_errors)
+            raise Bolt::RunFailure.new(set.error_set, 'apply_prep') unless set.ok
```

Comment:
In the facts stuff we make the "action" in the run_failure `run_task`. I think this is better but we should put a TODO on line 141 so we dont' forget to standardize when we do the tasks work.